### PR TITLE
Add support for Option.ofString.

### PR DESCRIPTION
**Pull request by isaacabraham**
An often-useful addition to the Option module for working with Strings.

### CODE REVIEW

**Comment by KevinRansom** [comment]
@isaacabraham Thanks, for this.  Because it adds an API to FSharp.Core it will be in the next language upgrade release.

**Comment by KevinRansom** [comment]
@isaacabraham 
because this PR adds an api the unit test baselines for public apis need updating.
They are there to ensure we don't sneak in any new APIs without realizing it :-)

Here is a commit from @neoeinstein that updates them:  https://github.com/Microsoft/visualfsharp/commit/ca2ed9bfbec4710802f3782ce8840ec060618347


https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.portable7.fs
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.portable47.fs
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.portable78.fs
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.portable259.fs
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.net40.fs
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core.Unittests/SurfaceArea.coreclr.fs


Kevin

**Comment by KevinRansom** [comment]
@isaacabraham You might check on fslang if this is an API that has been discussed and approved.

https://github.com/fsharp/fslang-design

**Comment by isaacabraham** [comment]
@KevinRansom Happy to write the unit tests etc. - unfortunately master does not build locally so I'm actually writing at the moment without compiling, which makes anything more than the most basic changes really difficult.

**Comment by isaacabraham** [comment]
Regarding APIs / RFCs - completely happy for this PR to be shelved / rejected - it's a tiny one. I mostly wanted to include this as a way to at least try and get up and running with contributing to the project.

**Comment by vasily-kirichenko** [comment]
Isn't it too much code for a such primitive one line function?..

**Comment by isaacabraham** [comment]
@vasily-kirichenko as far as I understand it, all these things are required: -

* FSI signature file
* Unit tests (maybe I went a bit overboard with two ;-)
* Surface area files for F# Core (although I have no idea what this is for)

But yes, adding a simple one line function required changes in multiple places. I'm also looking at the changes to the String module now - it's a similar situation there.

File: src/fsharp/FSharp.Core/option.fs [hunk 1]
```diff
@@ -85,6 +85,10 @@ namespace Microsoft.FSharp.Core
         [<CompiledName("OfNullable")>]
         let ofNullable (value:System.Nullable<'T>) = if value.HasValue then Some value.Value else None
 
+        [<CompiledName("OfString")>]
+        let ofString (value:System.String) =
+            if System.String.IsNullOrWhiteSpace value then None else Some value
```
Comment (forki) [inline src/fsharp/FSharp.Core/option.fs]: I wonder if this is OK. I personally think semantics should only be ifnull

Comment (isaacabraham) [inline src/fsharp/FSharp.Core/option.fs]: Good question. In my personal experience we always use IsWhitespace, but perhaps that's just for what we've used in the past.

If it's just if null you may as well just use ``OfObj`` - this is designed to be a little more useful than that with regards to dealing with strings specifically.

**Comment by forki** [comment]
Ok maybe isnullorempty, but whitespace is often signal.

Am 28.05.2017 12:53 nachm. schrieb "Isaac Abraham" <notifications@github.com

**Comment by isaacabraham** [comment]
We could either: -

* Make a second method
* Parameterise this one with a DU
* Just pick the one that is most popular; the other one would need people to fallback to the BCL version (not a terrible outcome I guess). Perhaps @cartermp has some nice stats on whether `IsNullOrEmpty` is used more than `IsNullOrWhitespace` (although the latter was only introduced in either .NET 3.5 or 4 I think?)

File: src/fsharp/FSharp.Core/option.fs (hunk 1 above)
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/option.fs]: I think IsNullOrWhiteSpace feels more right.  It feels more natural to map an empty string to None than Some ""

**Comment by cartermp** [comment]
@isaacabraham `IsNullOrEmpty` is used in about twice as many projects as `IsNullOrWhitespace` according to APIPort telemetry.  Website is (embarrassingly) unsecure if you want to see it yourself: https://apisof.net/catalog/System.String.IsNullOrEmpty(String)  (I only did so in a VM)

However, I think mapping `IsNullOrWhitespace` to `None` is better here.  If it weren't, I suspect there would be a lot of code like this:

```fsharp
match result with
| None -> bar()
| Some s -> if s = "" then foo() else bar()
```

File: src/fsharp/FSharp.Core/option.fs (hunk 1 above)
Comment (isaacabraham) [inline src/fsharp/FSharp.Core/option.fs]: I would say you could easily create that yourself by composing String.trim with the upcoming additions to the String module.

**Comment by dsyme** [comment]
@isaacabraham Yep, needs a suggestion and an RFC :)

https://github.com/fsharp/fslang-design/tree/master/RFCs

Thanks!

**Comment by dsyme** [comment]
> Regarding APIs / RFCs - completely happy for this PR to be shelved / rejected - it's a tiny one. I mostly wanted to include this as a way to at least try and get up and running with contributing to the project.

Understood.   Quick feedback on the design:

- I think we'd minimally need to rename the function (I had assumed by the name it was a parser for option values).   
- Nothing in FSharp.Core currently does anything special with whitespace.  
- Also one might want expect symmetry with any future ``Array.ofString`` or ``List.ofString`` and I don't think they would trim whitespace...

But that could be discussed in an RFC.  (Or just close this if you've extracted the learning value from the exercise :) )

**Comment by isaacabraham** [comment]
Created the RFC here https://github.com/fsharp/fslang-design/pull/207

**Comment by dsyme** [comment]
I'm closing this as the suggestion got [a considerable amount of negative feedback](https://github.com/fsharp/fslang-design/issues/224), and as a result the [RFC](https://github.com/fsharp/fslang-design/blob/master/drafts/FSS-0001-option-of-string.md) hasn't proceeded beyond draft.

**Comment by xperiandri** [comment]
Well if name it right it would be nice.
`Option.ofEmptyString`
`Option.ofWhiteSpaceString`

**Pull request closed without merging**